### DISCUSSION TITLE
Export / resampling: filename improvements

### DISCUSSION THREAD

**Discussion by nilscreates** [SynthstromAudible/DelugeFirmware#821] [Ideas]
It would be super helpful to make the filename of resampled tracks more intuitive / meaningful. Instead of just a sequential number, it could be something like the project name with sequential suffix, or even more detailed like _filename = Projectname-trackname-barlength_ as suggested by @sichtbeton on the discord.

**Comment by ok-reza** [comment]
only current benefit to the current system is that it is ordered by date, which is handy to find the latest one easily. 
what if you are resampling and haven't saved the song yet or that song title changes. i think that would create some confusion. whether that's still better than current implementation or not, i dont know.

more context is interesting (bar length) however there will be wild use cases with resampling, where tempo is constantly changing while resampling, so there will be edge cases to ponder on too.

**Reply by sichtbeton** [reply]
Both things could be true at once: have a date in the filename for ordering and also the name of the track for recognition. Bar length was only in there to have a measure to put into when trying to use it in a audio clip as you currently cannot have an unquantised audio clip (and need to give the exact bar length for it to play nicely). For the edge case of mixed bpm you could just display “misc.” or something along these lines.

**Comment by mungoooo** [comment]
I was just typing out this very request but I see I'm not the first to think of it! This feature would be INCREDIBLY useful for me as I'm constantly looking through past resampled audio and never remembering where the files came from.

Would it not be possible to have this option added as an alternate default in the deluge settings? I'd love to implement it myself, but have no experience in C++ (though I'd be more than willing to jump in to try get this done if it seems feasible!!).

**Reply by seangoodvibes** [reply]
We've added this for the audio / stem exports so someone could just leverage that logic and extend it to resampling

Or just point the resampling code to the audio export class to generate the file name

**Reply by seangoodvibes** [reply]
https://github.com/SynthstromAudible/DelugeFirmware/blob/main/src/deluge/processing/stem_export/stem_export.cpp#L1063

**Reply by mungoooo** [reply]
Thanks for the help. I managed to find a slightly simpler solution I think. Idk how to do pull requests and such, so I'll just copy the code snippet below in case anyone is interested. 

It just grabs the song name and concatenates it to the resample file name. "REC12345" --> "REC12345-Songname".

From audio_file_manager.cpp:

```
	// Build filename: REC#####[-SongName].WAV
	error = filePath->concatenate("/REC");
	if (error != Error::NONE) {
		return error;
	}
	error = filePath->concatenateInt(highestUsedAudioRecordingNumber[folderID], 5);
	if (error != Error::NONE) {
		return error;
	}
	if (!songName->isEmpty()) {
		error = filePath->concatenate("-");
		if (error != Error::NONE) {
			return error;
		}
		error = filePath->concatenate(songName->get());
		if (error != Error::NONE) {
			return error;
		}
	}
	error = filePath->concatenate(".WAV");
	if (error != Error::NONE) {
		return error;
	}

	if (doingTempFolder) {
		error =
		    tempFilePathForRecording->concatenate(&filePath->get()[strlen(audioRecordingFolderNames[folderID])]);
		if (error != Error::NONE) {
			return error;
		}
	}

	*getNumber = highestUsedAudioRecordingNumber[folderID];

	return Error::NONE;
}
```
